hax/hax/motr/__init__.py

```diff
@@ -110,6 +112,36 @@ def _entrypoint_request_cb(self, reply_context: Any, req_id: Any,
                   " '{}', process fid = {}".format(remote_rpc_endpoint,
                                                    str(process_fid)) +
                   ' The request will be processed in another thread.')
+        try:
+            cns = ConsulUtil()
+            if is_first_request and (not cns.is_proc_client(process_fid)):
+                # This is the first start of this process or the process has
+                # restarted.
+                # Let everyone know that the process has restarted so that
+                # they can re-establish their connections with the process.
+                # Disconnect all the halinks this process is any.
+                # Motr clients are filtered out as they are the initiators of
+                # the rpc connections to motr ioservices. Presently there's no
+                # need identified to report motr client restarts, Consul will
+                # anyway detect the failure and report the same so we exclude
+                # reporting the same during their first entrypoint request.
+                # But we need to do it for motr server processes.
+                q: Queue = Queue(1)
+                LOG.debug('first entrypoint request, broadcasting FAILED')
+                self.queue.put(
+                    BroadcastHAStates(
+                        states=[HAState(fid=process_fid,
+                                        status=ServiceHealth.FAILED)],
+                        reply_to=q))
+                ids: List[MessageId] = q.get()
+                LOG.debug('waiting for broadcast of %s for ep: %s',
+                          ids, remote_rpc_endpoint)
+                self.herald.wait_for_all(HaLinkMessagePromise(ids))
```

Comment:
Is it safe to invoke `wait_for_all` in Motr locality thread (note that we're still in Motr-land thread)? It will be blocked by this call for potentially a long time.

More than that, please note that this particular function is invoked from `hax.c` where GIL is locked. It is locked until `def _entrypoint_request_cb` exits. That means that NO other Python thread will be active until that moment. The whole hax application will freeze until `wait_for_all()` gets unblocked - that will happen from another Motr threads, so there is no deadlock. But freezing is still a bad thing in my opinion.

Suggestion: move this logic to ConsumerThread.

Reply:
Agree but we want to wait until the FAILED notifications are delivered to make sure there's no race between entrypoint response and FAILED notification.
Actually there's a bug here (known). Ideally this FAILED broadcast needs to be to entire cluster and not just local. I have an implementation for this in works but its not working cleanly yet.
Presently we do not have a way to block asynchronously until the FAILED notifications are delivered. I am trying a callback mechanism along with using BQ. I wanted to discuss this with you actually.

Reply:
But for now I want to get the basic changes in then fix the global broadcast case separately.

Reply:
I don't quite understand, why can't we just move all this processing to another thread. If `self.herald.wait_for_all` is invoked from a Pythonic thread (not Motr one), the problem of the application freeze will disappear.

Reply:
Yes we can, then we need to move posting entrypoint request where ever we move FAILED notification and wait_for_all. We can do this by introducing a new instance type that will also post entrypoint request after waiting_for_all deliveries.
But if we move this to Consumer thread then Consumer thread will be blocked, right?

Reply:
Yes, that's true. But in that case ConsumerThread will be the **only** thread being blocked at that point.

In your code, since the thread is external, nobody will release GIL till Condition gets triggered. That means that all Python threads are frozen until Motr finishes its dirty deals 😁

Reply:
Sorry but I need to understand how come GIL is released when we wait in ConsumerThread and why it is not released when we wait in motr thread.
But overall I see one problem with this, motr has a thread per locality (locality is per cpu core) and every type of operation is assigned a locality. In this case lets consider 2 operations hlink broadcast and entrypoint request. Now, in motr hlink operation and entrypoint operation both run in the same locality and thus share the thread. So if one operation blocks (entrypoint) other operation will get affected (hlink broadcast), I was having this suspicion which is causing the delay in overall motr process startup. From this perspective either we need to fix this in motr or in hax. I would prefer hax. So yes we need to release the motr thread.
Let me try something. But let's talk when you have time.

Reply:
I was thinking this is how it works,
```
The python interpreter creates a new process and spawns the threads
When thread-1 starts running, it will first acquire the GIL and lock it.
If thread-2 wants to execute now, it will have to wait for the GIL to be released even if another processor is free.
Now, suppose thread-1 is waiting for an I/O operation. At this time, it will release the GIL, and thread-2 will acquire it.
After completing the I/O ops, if thread-1 wants to execute now, it will again have to wait for the GIL to be released by thread-2.
```
is it that motr thread will not release the GIL because its an external thread?

Reply:
By the way from the logs seem things are moving in hax while the entrypoint request thread is blocked,
```
Dec 03 20:43:14 ssc-vm-c-0552.colo.seagate.com hare-hax[14000]: 2020-12-03 20:43:14,005 [DEBUG] {Dummy-1} Blocking until HaLinkMessagePromise{MessageId(0x7f202401e710, 9), MDec 03 20:43:14 ssc-vm-c-0552.colo.seagate.com hare-hax[14000]: 2020-12-03 20:43:14,016 [DEBUG] {qconsumer} The repeatable call succeeded: update_process_status
Dec 03 20:43:14 ssc-vm-c-0552.colo.seagate.com hare-hax[14000]: 2020-12-03 20:43:14,016 [DEBUG] {qconsumer} Waiting for the next message
Dec 03 20:43:17 ssc-vm-c-0552.colo.seagate.com hare-hax[14000]: 2020-12-03 20:43:17,142 [DEBUG] {fs-stats-updater} Attempting to invoke the repeatable call: _ioservices_runnDec 03 20:43:17 ssc-vm-c-0552.colo.seagate.com hare-hax[14000]: 2020-12-03 20:43:17,147 [DEBUG] {fs-stats-updater} Services "confd" received: [{'ID': 'f6dc1035-d338-339c-83fDec 03 20:43:17 ssc-vm-c-0552.colo.seagate.com hare-hax[14000]: 2020-12-03 20:43:17,149 [DEBUG] {fs-stats-updater} Raw value from KV: {'LockIndex': 0, 'Key': 'processes/0x72Dec 03 20:43:17 ssc-vm-c-0552.colo.seagate.com hare-hax[14000]: 2020-12-03 20:43:17,151 [DEBUG] {fs-stats-updater} Raw value from KV: {'LockIndex': 0, 'Key': 'processes/0x72Dec 03 20:43:17 ssc-vm-c-0552.colo.seagate.com hare-hax[14000]: 2020-12-03 20:43:17,152 [DEBUG] {fs-stats-updater} Services "ios" received: [{'ID': 'f6dc1035-d338-339c-83f9-Dec 03 20:43:17 ssc-vm-c-0552.colo.seagate.com hare-hax[14000]: 2020-12-03 20:43:17,154 [DEBUG] {fs-stats-updater} Raw value from KV: None
Dec 03 20:43:17 ssc-vm-c-0552.colo.seagate.com hare-hax[14000]: 2020-12-03 20:43:17,155 [DEBUG] {fs-stats-updater} Raw value from KV: None
```
No?

Reply:
You won 👍 

It seems like I was too pessimistic on external non-python threads. Please feel free to ignore the request in this conversation.